dynamic/binary_test.go

```diff
@@ -131,4 +131,13 @@ func TestBinaryUnknownFields(t *testing.T) {
 
 func binaryTranslationParty(t *testing.T, msg proto.Message) {
 	doTranslationParty(t, msg, proto.Marshal, proto.Unmarshal, (*Message).Marshal, (*Message).Unmarshal)
+
+	// Declare a function that has the same interface as (*Message.Marshal) but uses
+	// MarshalAppend internally so we can reuse the translation party tests to verify
+	// the behavior of MarshalAppend in addition to Marshal.
+	marshalAppend := func(m *Message) ([]byte, error) {
+		b := []byte{}
+		return (*Message).MarshalAppend(m, b)
```

Comment:
This is a strange way to invoke the method. Why not just `m.MarshalAppend(b)`?

Also, I think it's worth doing some more assertions here, to verify that this is doing the desired thing.

For example, instead of passing in a zero-capacity slice, pass in one that should fit more/all of the marshaled message. When you get back the result, you can check if its length is <= the capacity of the slice. If so, then it should have re-used that capacity and done no allocations, so you can then add a check to verify that they actually point to the same array. You can do that check using reflection:

```go
origPtr := reflect.ValueOf(b).Pointer()
resultPtr := reflect.ValueOf(result).Pointer()
testutil.Eq(t, origPtr, resultPtr)
```

Reply:
You might also have this function invoke `MarshalAppend` twice: once with an empty buffer and again with actual data. Then you can verify that the contents of the supplied buffer are retained as a prefix of the marshaled data. Does that make sense?

Reply:
Great feedback, will implement. We do something similar to the reflection pointer comparison in M3DB: https://github.com/m3db/m3/blob/7451870e986a61083d47d0c1b83a32035e6fb098/src/x/test/util.go#L30

Reply:
Ok I updated the tests